### PR TITLE
Insert Links by default in Navigation block

### PR DESCRIPTION
**Pull request by tellthemachines**
## Description

First steps to address #34041:

Changes Navigation block appender to insert Navigation Links by default, unless there are other types of blocks in the Navigation block already. 

This is a rough draft so we can have a play and see how it feels. 

One thing I'm not sure about is how it should behave when we "Add all pages". Due to the Page List being a different block type, should we show the quick inserter? Or should we treat Page List as if it were just a bunch of Navigation Links?

~~The other question is how to address submenu: when a Submenu block is added, the quick inserter will appear because Submenus are a different block type. But because they are often part of simple navs, should we consider them the same as Navigation Links?~~

~~Likewise, it might be good to insert links by default inside submenus too.~~

Updated to insert links by default when submenus are present (and inside submenus).

## How has this been tested?

* Add a Navigation block, choose "Start Empty" and click the plus button. A Navigation Link block should be inserted straightway. 
* From the top inserter, insert a different type of block, e.g. Search. Clicking the plus button should now show the quick inserter.
* In the Navigation screen, check that Navigation Link blocks are inserted straightway.

## Types of changes

New feature (non-breaking change which adds functionality)

This is more of an enhancement than a new feature, but 🤷 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
Your stamina to take one difficult ticket after another is impressive. Thank you. It's already looking good:

nav

As is, this feels very validating towards the concept as a whole. Working towards the second half, transforms, it's already becoming clear that showing the 3 most recent items published in the page selector, requiring search for anything else, isn't that useful. This is almost certainly not something that should be solved in this PR, and there are [some mockups in this ticket](https://github.com/WordPress/gutenberg/issues/30170#issuecomment-920719539), but it feels like a worthwhile next step.

Let me know if/how I can help with this one.

**Comment by gwwar** [comment]
Thanks for looking into this @tellthemachines ! I gave this a try, and one thing I quite like about how the buttons container works currently is that it quickly lets us add our placeholders before we need to fill in the details.

In the context of the navigation editor, this branch lets us add items pretty quickly, but I got pretty distracted by the link search that auto opens. If we're trying to quickly build a submenu, this makes it a little more apparent that it can get in the way if I'm building a menu in that way.

If we prevent the link search from auto opening, a natural thing to do might also be to add an option to the link search to open the quick inserter if we're not interested in a link:



https://user-images.githubusercontent.com/1270189/134069576-ffcede12-4cac-4b3a-ac64-ade7840e2331.mp4

Note that if this PR ends up being viable, we can also use the method here to improve how social icons inserts items in a follow up✨

**Comment by gwwar** [comment]
> Let's try making it insert a Page Item by default

If we want page links, we'll need to add the following attributes to make sure we get the Select Page prompt:
```
<!-- wp:navigation-link {"type":"page","kind":"post-type" } /-->
```
CleanShot 2021-09-20 at 13 46 31@2x

**Comment by tellthemachines** [comment]
> > Let's try making it insert a Page Item by default
> 
> If we want page links, we'll need to add the following attributes to make sure we get the Select Page prompt:

Updated to Page items by default, but I now realise that this complicates the flow for inserting other types of links: searching for e.g a category name won't show any results. In optimising for Pages we risk making the general experience worse instead of better 😕
I'd love some more thoughts on this!

> In the context of the navigation editor, this branch lets us add items pretty quickly, but I got pretty distracted by the link search that auto opens. If we're trying to quickly build a submenu, this makes it a little more apparent that it can get in the way if I'm building a menu in that way.

The link control auto-opening is what happens currently when adding a Link block, so changing that isn't in scope for this PR. In any case, I'm not convinced we should change it. Nav links are different to buttons in that they are most often based on existing pages or posts, so it doesn't make sense to add a text placeholder for them and go back later to add the actual link. That would just be twice the amount of work 😅 

I'm going to mark this one ready for review, but I'll hold off on updating the e2e tests until I have some more feedback on Page items by default, because going with that will force us to change quite a few things.

File: packages/block-library/src/navigation/edit.js
Comment (tellthemachines) [inline packages/block-library/src/navigation/edit.js]: This is memoized to stop the directInsertValue array from getting recreated on every render, and causing an infinite loop because it changes innerBlocks, which in turn the Navigation function depends on to calculate the directInsertValue 😖 

I can't think of a better way to do it though. I think this logic should reside in the block, because it might vary a lot - Submenus are set to always use it (at least until we sort out mega menus), and conceivably there could be blocks using it depending on their nesting status or some other condition.

**Comment by talldan** [comment]
I have mixed feelings. Inserting pages is fast and easy, and is probably what most users will want to do, so that's great.

But it also feels like the other options are very hidden and can only be added in a very indirect way. There are probably ways to mitigate this. Perhaps by improving the Link UI (I think @jasmussen did some explorations) or I think @mtias mentioned making the slash inserter usable (though you can't just type into a Page block, a link has to be selected first) alongside other ideas in the issue - https://github.com/WordPress/gutenberg/issues/34041.

File: packages/block-library/src/navigation-submenu/edit.js
Comment (talldan) [inline packages/block-library/src/navigation-submenu/edit.js]: There are existing mechanisms in variations and blocks to define whether they can be used in the inserter. Introducing some level of granularity there might be an alternative to this prop.

e.g. 
```
"supports": {
    "inserter": "__experimentalQuick" | "__experimentalGlobal" | true | false
}
```

Comment (talldan) [reply packages/block-library/src/navigation-submenu/edit.js]: I remember @gwwar also looked into a default block for the navigation block, so going back to that might also be an option.

**Comment by getdave** [comment]
> Perhaps by improving the Link UI (I think @jasmussen did some explorations) 

We're looking at this in https://github.com/WordPress/gutenberg/pull/33849#issuecomment-920121414.

File: packages/block-library/src/navigation/edit.js
Comment (mtias) [inline packages/block-library/src/navigation/edit.js]: We should add a comment here explaining this better for the future

**Review by getdave — COMMENTED** [review]
Thank you for working on this.

I recognise the need for a lighter and faster experience for adding links.

My instinct is that this may not be the right solution. It makes adding Pages fast but it makes adding any other kind of link variation far more obscure.

One option we may not have considered is enabling some kind of "bulk addition" UI as a complementary alternative to the inline inserter. We're actually [looking into this at the moment in the Navigation Editor](https://github.com/WordPress/gutenberg/issues/31667) and I wonder whether perhaps this ability to insert any type of item in bulk without triggering the Link UI (which I'm happy to work on or make configurable btw!) would also benefit the block?

I also think it would be worth exploring the scenario where:

* clicking plus auto-inserts a `core/navigation` link block without triggering the Link UI.
* you can repeatedly add navigation link items _without_ multiple clicks to close the link UI.
* once you're done with the "_adding_ items" bit you can click on any item to "configure it" (we could show a visual indicator this is required).

I think this is what @gwwar suggested above. That would be a nicer flow in my opinion. It would have two (conceptual) "modes":

* insertion mode (I'm adding items).
* configuration mode (I'm setting where the items link to).

Thanks for working on this. ~I'm going to go away now and look at home to make the auto-appearing link UI popup configurable~. I did an experiment https://github.com/WordPress/gutenberg/pull/35001

**Comment by draganescu** [comment]
I like this idea suggested by Dave

> - clicking plus auto-inserts a core/navigation link block without triggering the Link UI.
> - you can repeatedly add navigation link items without multiple clicks to close the link UI.
> - once you're done with the "adding items" bit you can click on any item to "configure it" (we could show a visual indicator this is required).

**Comment by priethor** [comment]
> once you're done with the "adding items" bit you can click on any item to "configure it" (we could show a visual indicator this is required).

As I commented in #35001 (comment), I feel switching the insertion flow and asking users to first add all links, then set them, is a hard ask and not the smoothest insertion flow.

If we want to go for a bulk-add hybrid approach instead, how about we explore adding a different appender to "insert another" of the same kind? This would allow, for example, inserting successive page links right after adding a first one, so that creating simple link menus is very straightforward but without obscuring the ability to add other types of blocks.

**Comment by adamziel** [comment]
The way I understand the discussion is this:

* Inserting links without asking whether that's a page, post, tags, category etc feels smooth – that's great
* But if we insert a page link, it hides the posts, tags, and categories links from the user – that's bad

Can we eat a cake and have it too? Imagine a link UI that shows anything that matches: pages, posts, tags, categories, you name it:

Zrzut ekranu 2021-09-21 o 17 09 38

Once you choose a tag, it becomes a tag link. If you edit it, it could still show all results for consistent experience – this way you could easily change it into a page link. Would that work, or am I missing something here?

File: packages/block-library/src/navigation-submenu/edit.js
Comment (gwwar) [reply packages/block-library/src/navigation-submenu/edit.js]: The related PR was here, where we explored adding a default block per container (vs one for the entire editor). https://github.com/WordPress/gutenberg/pull/29490.

If I recall we ran into some of the same issues, where we made a guess as to what folks wanted to insert, but transforming blocks after that was more difficult.

Comment (gwwar) [reply packages/block-library/src/navigation-submenu/edit.js]: > Let's highlight the default block better in the quick inserter. 

🤔 From https://github.com/WordPress/gutenberg/issues/34041 it might be worth trying an `__experimentalDefaultBlock` block list setting again.

I'll try tidying that experimental branch to see if its viable.

Comment (tellthemachines) [reply packages/block-library/src/navigation-submenu/edit.js]: > There are existing mechanisms in variations and blocks to define whether they can be used in the inserter. Introducing some level of granularity there might be an alternative to this prop.

I don't think a setting in the block.json will work in this case, because we need to change it based on the block's innerBlocks. If any of the innerBlocks are _not_ Nav Links, the quick inserter should appear instead of directly inserting a Link.

For the same reason, I'm not sure we'd want to leverage @gwwar 's `__experimentalDefaultBlock`, because the Insert before and Insert after buttons would be tied to its value. If we use `__experimentalDefaultBlock`, as it stands in #29490, to tell the block inserter when we want a default block to be inserted directly, we'll end up disabling Insert before and Insert after whenever the Navigation contains anything other than Links, which is quite unexpected.

Independently of whether we do enable Insert before and Insert after on all block containers in #29490 (which I think we should! it makes total sense) we still want to make it easier to insert Nav Links inside the Nav block, and a workflow that forces us to look for a deeply hidden option in the block toolbar doesn't satisfy that requirement. 

In order to keep `useInnerBlocksProps` simple it would be nice if we could leverage the same prop for both use cases, but I'm not sure it will be possible. Am keen to know if anyone has other ideas around this though!

Comment (talldan) [reply packages/block-library/src/navigation-submenu/edit.js]: That part of it is pretty complicated. There's still this part of the issue that indicates a way to designate a default block within the nav block will be needed:

> Let's highlight the default block better in the quick inserter. It should always come in first. A user should not have to think when they open a quick inserter if they are not looking for a specific block.

I don't know that the current `directInsert` prop will work well with that.

Maybe an option is to look at using `renderAppender` to build out a lot of the custom insertion behavior first before formalising it into props and configuration.

**Comment by talldan** [comment]
I second @adamziel's suggestion, but with a slight twist.

I think it could work by:
- Don't show navigation link variations (page, post, tag) in the quick inserter, just the base 'Link' block. The link block is inserted by default when clicking `+`, a bit like we're exploring here. (Reasoning: the quick inserter and main inserter are overwhelming with too many options, particularly when a site has lots of custom post types or taxonomies.)
- The URL picker shows pages by default for the link block, but also has a way to switch to other types or all types. (Reasoning: pages are what a user will usually want to insert)
- The selected link type determines which variation is used (this part should already work, you can see it by selecting a Custom Link, and then choosing a Page).

**Comment by tellthemachines** [comment]
Thanks for the feedback, everyone!

> But it also feels like the other options are very hidden and can only be added in a very indirect way. There are probably ways to mitigate this. Perhaps by improving the Link UI

Leveraging transforms to add visibility to other options is already being explored in #34978. It would be good to look at using the slash inserter too, but that should be a separate PR.

> My instinct is that this may not be the right solution. It makes adding Pages fast but it makes adding any other kind of link variation far more obscure.

> Can we eat a cake and have it too? Imagine a link UI that shows anything that matches: pages, posts, tags, categories, you name it:

I feel the same about making the link control Page-specific. I initially made it generic, and then changed it to Pages as the brief in #34041 mentions them specifically. It's super easy to change back though! 

> One option we may not have considered is enabling some kind of "bulk addition" UI as a complementary alternative to the inline inserter.

That could work, but I'm not sure it would be best as part of the link control. It would be unexpected to add a single block and then have that block generate a bunch of siblings. Maybe a different kind of inserter, or a toolbar button opening a modal or something? It would need some design input in any case 😅 

> clicking plus auto-inserts a core/navigation link block without triggering the Link UI.
> you can repeatedly add navigation link items without multiple clicks to close the link UI.

I'm really not sold on this; it complicates the process with no obvious benefit. I may be missing something, but placeholders don't feel like a useful tool for nav-building.

> The URL picker shows pages by default for the link block, but also has a way to switch to other types or all types. (Reasoning: pages are what a user will usually want to insert)

It would be good if the autocomplete prioritised showing Page results, but searching brought up all other post types/taxonomies. That should probably be handled in a separate PR though.

**Comment by tellthemachines** [comment]
Ok folks, I've updated to insert a generic Link instead of a Page link. Let me know what you think!

**Comment by talldan** [comment]
> It would be good if the autocomplete prioritised showing Page results, but searching brought up all other post types/taxonomies. That should probably be handled in a separate PR though.

It's a very small UI though, you only see 3 search results at a time, so it could make it very difficult to add anything other than a page without frustrating scrolling.

I mentioned it to @javierarce in another issue, but this area feels very much like an area that could do with UX research. It doesn't help that we develop these features on dev instances that typically only have a few search items. We really need to see how this works on production instances that have thousands of things to search through.

**Comment by getdave** [comment]
> [@getdave]: One option we may not have considered is enabling some kind of "bulk addition" UI as a complementary alternative to the inline inserter.
> [@tellthemachines]: That could work, but I'm not sure it would be best as part of the link control. It would be unexpected to add a single block and then have that block generate a bunch of siblings. Maybe a different kind of inserter, or a toolbar button opening a modal or something? It would need some design input in any case 😅

For context, [the Issue I linked up has designs for exactly that](https://github.com/WordPress/gutenberg/issues/31667). We shouldn't be making the Link UI any more complex. Rather we have another option for inserting blocks that _complements_ the Link UI.

👉 Please see the [Bulk Addition UI Issue](https://github.com/WordPress/gutenberg/issues/31667) for more info.

> Don't show navigation link variations 

100% agree here. Users don't need to understand all the differences between Post, Page upfront _before_ they see results. Let's simplify.

> The selected link type determines which variation is used

We need to double check that this is working right now.

> It's a very small UI though, you only see 3 search results at a time, so it could make it very difficult to add anything other than a page without frustrating scrolling.

I [bumped the Nav Editor up to 10 results at a time](https://github.com/WordPress/gutenberg/pull/34808). We could do the same **_but..._**

We could _also_ have a "Browse All" button which opens a new "Content" sidebar panel _pre-filtered_ with search results. This panel could closely mirror the designs for the [Bulk Addition UI for the Nav Editor](https://github.com/WordPress/gutenberg/issues/31667).

File: packages/block-library/src/navigation-submenu/edit.js
Comment (gwwar) [reply packages/block-library/src/navigation-submenu/edit.js]: I'm not sure if `__experimentalDefaultBlock` is viable, but the idea of a default block other than a global one for the editor may still be useful. (The buttons block inserter cheats a bit since only one child item is allowed and doesn't apply well to other situations). 

I mostly just rebased the branch yesterday and resolved conflicts. I'll ping y'all if I think there might be viable flow to look at.

**Comment by gwwar** [comment]
> Don't show navigation link variations

For this case, the quick inserter was probably not a good fit for this in the first place, though they are useful for narrowing down what type of items to add (in search). I can spin up a draft to remove the variations from the quick inserter to let folks play with it. Edit: see https://github.com/WordPress/gutenberg/pull/35056

Was https://github.com/WordPress/gutenberg/issues/31667 ready to be worked on? The additional space for bulk addition looks much more comfortable in the sidebar vs popover.

**Comment by getdave** [comment]
> Was #31667 ready to be worked on? The additional space for bulk addition looks much more comfortable in the sidebar vs popover.

I've started a PR locally but been sidetracked. I'll push up where I am tomorrow.

File: packages/block-library/src/navigation-submenu/edit.js
Comment (tellthemachines) [reply packages/block-library/src/navigation-submenu/edit.js]: > Maybe an option is to look at using renderAppender to build out a lot of the custom insertion behavior first before formalising it into props and configuration.

Not sure if I'm missing something, but `renderAppender` seems to be used only for determining what the appender looks like. It doesn't feel right to mix that with its behaviour (whether it should display quick inserter or insert a block directly), because we'll want to be able to set both independently.

> There's still this part of the issue that indicates a way to designate a default block within the nav block will be needed

This is a good point. I'm thinking we could have two props: a `defaultBlock` array that specifies the default block name and any relevant attributes, and a `directInsert` boolean that specifies whether the default block, if it exists, should be inserted directly or not. That way `defaultBlock` can be used to specify the default for the quick inserter, as well as for insert before and after. If I'm not mistaken `defaultBlock` will need to be passed to `useInnerBlockProps` for all of these scenarios.

Comment (tellthemachines) [reply packages/block-library/src/navigation-submenu/edit.js]: @gwwar I'm going to steal the`__experimentalDefaultBlock` prop and give it a go in this PR 😂

Comment (gwwar) [reply packages/block-library/src/navigation-submenu/edit.js]: Go for it ✨

**Comment by tellthemachines** [comment]
With #34978 merged, link variations being looked at in #35056 and #35104, and the bulk addition UI discussed in  #31667, the pieces of this puzzle are starting to come together 🚀 

I changed the implementation here to separate out the default block type and whether the default block should be inserted directly into two distinct props. That way we'll be able to leverage the default block type for other uses, as discussed above. 

It would be good to get a further round of review before tackling the e2e tests, as a bunch of changes will need to be made there, but they depend a lot on what we settle on doing here 😅

**Comment by adamziel** [comment]
@tellthemachines The changes make sense to me. If I understand correctly, this makes it possible to only use direct insertion for some blocks lists, but not for the other? That's nice and could come handy when mixing different contexts in the same editor, or using the same block in different editors.

File: packages/block-editor/src/components/inserter/index.js
Comment (gwwar) [inline packages/block-editor/src/components/inserter/index.js:289]: I think the default block can work as an array or object. It may help to give a few more hints on the data structure with something like:

```jsx
const [ name, attributes, innerBlocks ] = defaultBlock;
createBlock( name, attributes, innerBlocks );
```

File: packages/block-library/src/navigation/edit.js
Comment (gwwar) [inline packages/block-library/src/navigation/edit.js]: Have we considered allowing `__experimentalDirectInsert` to be a match function which returns true or false?  It doesn't look like we need any other information specific to the render function. One benefit is that we can then make it a constant and can define it next to the default block.

For example it might look like:

```jsx
const DIRECT_INSERT = ( block ) => { 
    return block.innerBlocks.every( ( { name } ) => name === 'core/navigation-link' || name === 'core/navigation-submenu' );
};
```

And in the selector we can update it to call the match function.

**Review by gwwar — COMMENTED** [review]
This is looking good so far @tellthemachines! 💖 The interactions feel good to me in both the post and navigation editor. 

What we need next is updated tests and maybe a few more eyes on the new API. I left a few suggestions, but I'm curious what others think too.

File: packages/block-library/src/navigation/edit.js
Comment (tellthemachines) [reply packages/block-library/src/navigation/edit.js]: Oooh nice! I'll try it out.

Comment (tellthemachines) [reply packages/block-library/src/navigation/edit.js]: Hmm, this doesn't work well when we either add or remove other block types, because the selector won't re-run with every block update, so it won't know when the content has changed. Whereas if we keep the logic inside the edit function it updates with every change in the inner blocks.

File: packages/block-editor/src/components/inserter/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/inserter/index.js:289]: Can you think of a scenario where we'd need defaultBlock to be an object? In the existing use cases (both this and your other PR) we're using it to store values to pass to `createBlock` so it makes more sense as an array. 

In this PR, it's defined as an array in the jsdocs for the parameters passed to `useNestedSettingsUpdate` and the return value of `__experimentalGetDirectInsertBlock`. We'd need to change those if we wanted to support an object too.

**Comment by tellthemachines** [comment]
> If I understand correctly, this makes it possible to only use direct insertion for some blocks lists, but not for the other? That's nice and could come handy when mixing different contexts in the same editor, or using the same block in different editors.

Direct insertion depends on the block for now, though conceivably we could change it to be dependent on context if necessary. Right now it isn't because we want the same behaviour in all the editors.

> What we need next is updated tests and maybe a few more eyes on the new API.

Thanks for the review @gwwar ! I've updated the failing tests and added a new test to check the quick inserter still works when non-link blocks are present. Would welcome more feedback on the API!

File: packages/block-editor/src/components/inserter/index.js
Comment (adamziel) [reply packages/block-editor/src/components/inserter/index.js:289]: For me, a good reason to make it an object would be passing around a block instead of an array of arguments for `createBlock`. For example, this API could look like `const DEFAULT_BLOCK = createBlock( 'core/navigation-link' )`. One thing I don't like there is creating a uuid and a few other things we don't really need.

Maybe that could be a function then? For example `const DEFAULT_BLOCK_CREATOR = () => createBlock( 'core/navigation-link' )`

File: packages/block-library/src/navigation/edit.js
Comment (gwwar) [reply packages/block-library/src/navigation/edit.js]: Have you tried without `createSelector`? We should get proper results provided we recalculate whenever the block.innerBlocks reference updates. 

Edit: I see some earlier comments about an infinite loop. I'll 🔍 a bit. This is more of a nice to have if we can define this outside.

File: packages/block-editor/src/components/inserter/index.js
Comment (gwwar) [reply packages/block-editor/src/components/inserter/index.js:289]: 🤔 Good point, I suppose folks will need to use createBlock for complex innerBlocks and we'll want to avoid keeping around a static uuid for blocks. A function would make sense in that case, though we might want to pick a friendlier name. (I find that some folks get confused by the `creator` name).  Maybe `INSERT_DEFAULT_BLOCK` or something similar? 

Edit: Maybe a simple name like `getDefaultBlock`? One other piece of this is we'll likely need to modify the suggested items in the quick inserter as part of 34041. (The quick inserter is currently ordered by block frecency).

File: packages/block-library/src/navigation/edit.js
Comment (gwwar) [reply packages/block-library/src/navigation/edit.js]: The following works for me. (I didn't notice any infinite loops, but let me know if there's a specific case to test.)

Non-memo'd:

```jsx
export const __experimentalGetDirectInsertBlock = (
	state,
	rootClientId = null
) => {
	if ( ! rootClientId ) {
		return;
	}
	const settings = state.blockListSettings[ rootClientId ];
	const defaultBlock = settings?.__experimentalDefaultBlock;
	if ( ! defaultBlock ) {
		return;
	}
	const directInsert = settings?.__experimentalDirectInsert(
		getBlock( state, rootClientId )
	);
	if ( ! directInsert ) {
		return;
	}

	return defaultBlock;
};
```

Or memo'd

```jsx
export const __experimentalGetDirectInsertBlock = createSelector(
	( state, rootClientId = null ) => {
		if ( ! rootClientId ) {
			return;
		}
		const settings = state.blockListSettings[ rootClientId ];
		const defaultBlock = settings?.__experimentalDefaultBlock;
		if ( ! defaultBlock ) {
			return;
		}
		const directInsert = settings?.__experimentalDirectInsert(
			getBlock( state, rootClientId )
		);
		if ( ! directInsert ) {
			return;
		}

		return defaultBlock;
	},
	( state, rootClientId ) => [
		state.blockListSettings[ rootClientId ],
		state.blocks.tree[ rootClientId ],
	]
);
```

File: packages/block-editor/src/components/inserter/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/inserter/index.js:289]: Hmm, I'm not so sure about this. When allowedBlocks is in use, the block is created directly in the inserter. I'd prefer to emulate that logic: fetching a block type, and creating the block from that type in the inserter. Default block and allowed block shouldn't be too different, because they essentially serve the same purpose. The main difference is that only one default block may exist even if multiple blocks are allowed.

> I suppose folks will need to use createBlock for complex innerBlocks and we'll want to avoid keeping around a static uuid for blocks

I'm not sure I follow. We shouldn't have any reason to pass innerBlocks to the inserter, because it should always insert a brand new block.

File: packages/block-library/src/navigation/edit.js
Comment (tellthemachines) [reply packages/block-library/src/navigation/edit.js]: Oh I see, I was forgetting to add `state.blocks.tree[ rootClientId ]` to the dependencies 🤦 

I think we should also accept a boolean value though, to avoid having to do something like `() => true` when we want to _always_ insert the default block.

File: packages/block-editor/src/components/inserter/index.js
Comment (adamziel) [reply packages/block-editor/src/components/inserter/index.js:289]: > I suppose folks will need to use createBlock for complex innerBlocks and we'll want to avoid keeping around a static uuid for blocks

I think @gwwar referred to pre-populating that block with inner blocks. For example, that could be a pre-populated menu like this:

```js
const DEFAULT_BLOCK_CREATOR = () => createBlock( 'core/navigation-menu', {}, [
    createBlock( 'core/navigation-link', { title: "Home" }, [ ] ),
    createBlock( 'core/navigation-link', { title: "About" }, [ ] ),
] )
```

But as you say, I don't think it's a use-case that exists today.

> Default block and allowed block shouldn't be too different, because they essentially serve the same purpose. 

Hm I just thought of a completely different perspective here. Could we just make it a string to point out which supported block do we want to insert by default? It must be one of the supported blocks **and** we don't specify any attributes at the moment so maybe we don't need anything more? Or are we running in circles now?

File: packages/block-library/src/navigation/edit.js
Comment (gwwar) [reply packages/block-library/src/navigation/edit.js]: > I think we should also accept a boolean value though

That sounds reasonable to me 👍

File: packages/block-editor/src/components/inserter/index.js
Comment (gwwar) [reply packages/block-editor/src/components/inserter/index.js:289]: > I think @gwwar referred to pre-populating that block with inner blocks. For example, that could be a pre-populated menu like this:

That's right. A better example might be a container block that makes little sense without children, for example the buttons block.

Since we don't have an explicit case for passing though attributes or innerBlocks yet, I think it's fine deferring needing to pass through extra arguments until we need them. Whichever combination you prefer @tellthemachines

Comment (gwwar) [reply packages/block-editor/src/components/inserter/index.js:289]:
```jsx
const blockToInsert = directInsertBlock?.length
					? createBlock( ...directInsertBlock )
					: createBlock( allowedBlockType.name )
```

One thing to note is that if folks pass through innerBlocks as the third item in an array, it's possible to have innerBlocks with static uuids. It's an edge case, but two options I can think of:

1. clone the block
```jsx
insertBlock( cloneBlock( blockToInsert ), getInsertionIndex(), rootClientId );
```
2. Avoid passing through innerBlocks if we don't want to handle this case yet 
```jsx
const [ name, attributes ] = directInsertBlock;
createBlock( name, attributes )
```

**Review by gwwar — APPROVED** [review]
Thanks @tellthemachines I think we're in pretty good shape here. This tests well for me manually and the test updates make sense. ✨ 

We may want a +1 review from folks working on the Navigation Editor since this effectively makes it pretty difficult to add a link variation. The only flow I found for that was using the global inserter.  

Here's what I see. Note that the navigation crash at the end when hovering over a transform is probably from a different bug that has been fixed. (I'd recommend a rebase with trunk to verify).

https://user-images.githubusercontent.com/1270189/135313862-c9182156-2b66-4150-b5a2-823ad3dc8833.mp4

**Comment by tellthemachines** [comment]
> Note that the navigation crash at the end when hovering over a transform is probably from a different bug that has been fixed.

Yup, it's not happening after rebase.

**Review by talldan — APPROVED** [review]
LGTM, thanks for working on this.

**Comment by mtias** [comment]
Nice! This heuristic is going to be pretty important going forwards.

**Pull request merged**
